keanu-project/src/main/java/io/improbable/keanu/util/dot/WriteDot.java

```diff
@@ -0,0 +1,279 @@
+     * Read more about DOT format here: https://en.wikipedia.org/wiki/DOT_(graph_description_language)
+     *
+     * @param fileName name of the output file
+     * @param vertex vertex around which the graph will be visualised
+     * @param degree degree of connections to be visualised; for instance, if the degree is 1,
+     *               only connections between the vertex amd it's parents and children will be written out to the DOT file.
+     */
+    public static void outputDot(String fileName, Vertex vertex, int degree) {
+        try {
+            // Reset variables
+            idsToLabelStrings = new HashMap<>();
+            verticeIdsToConnectionInfo = new HashMap<>();
+
+            File outputFile = getOutputFile(fileName);
+
+            writer = new BufferedWriter(new FileWriter(outputFile));
+            writer.write("digraph BayesianNetwork {\n");
+
+            // Iterate over vertices and obtain the necessary label and conneciton info.
+            obtainGraphInfo(vertex, degree);
+
+            // Write out labels and connections.
+            outputConnections(verticeIdsToConnectionInfo.values());
+            outputLabels();
+
+            writer.write("}");
+            writer.close();
+        }
+        catch (IOException e) {
+            System.out.println("Encountered an issue writing to the specified file (" + fileName + "):");
+            e.printStackTrace();
+        }
+    }
+
+
+    // Make sure the output directory exists and that no files are being overwritten.
+    private static File getOutputFile(String fileName) {
+
+        File outputFile = new File(fileName);
+
+        // Make sure the output directory exists.
+        if (outputFile.getParentFile() != null && !outputFile.getParentFile().exists()){
+            outputFile.getParentFile().mkdirs();
+        }
+
+        // Add an index to the filename if a file with the specified name already exists.
+        if (outputFile.exists()) {
+            String baseName = FilenameUtils.getBaseName(fileName);
+            String extension = FilenameUtils.getExtension(fileName);
+            int counter = 1;
+            while(outputFile.exists())
+            {
+                outputFile = new File(outputFile.getParent(), baseName + "_" + (counter++) + "." + extension);
+            }
+        }
+        return outputFile;
+    }
+
+
+    /**
+     * Iterates over the vertices of the graph, storing information about vertex labels to display and connections between vertices.
+     *
+     * @param vertex vertex around which the graph will be visualised
+     * @param degree degree of connections to be visualised; for instance, if the degree is 1,
+     *               only connections between the vertex amd it's parents and children will be written out to the DOT file.
+     */
+    private static void obtainGraphInfo(Vertex vertex, int degree) {
+
+        HashSet<VertexId> processedVertices = new HashSet<>();
```

Comment:
Do you need to explicitly store this as a HashSet?  We usually just store variables as their "most generic" type (to not expose unnecessary implementation details to other bits of the code - ie over the long term you end up with something more maintainable).

Reply:
Oh, do you mean `Set<VertexId> processedVertices = new HashSet<>();`?